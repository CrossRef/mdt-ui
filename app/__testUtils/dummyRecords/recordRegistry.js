import defaultArticle from './defaultArticle'
import fullFormArticle from './fullFormArticle'


export default {
  [defaultArticle.articleDoi]: defaultArticle.articleJson,
  [defaultArticle.publicationDoi]: defaultArticle.publicationJson,

  [fullFormArticle.articleDoi]: fullFormArticle.articleJson,
  [fullFormArticle.publicationDoi]: fullFormArticle.publicationJson

}
