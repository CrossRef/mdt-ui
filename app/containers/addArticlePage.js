import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import is from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import {browserHistory} from 'react-router'
import $ from 'jquery'

import defaultState from '../components/AddArticlePage/defaultState'
import { controlModal, getPublications, editForm, deleteCard, clearForm, cartUpdate } from '../actions/application'
import {finishUpdate} from '../utilities/helpers'
import {cardNamesArray} from '../utilities/crossmarkHelpers'
import AddArticleView from '../components/AddArticlePage/addArticleView'
import {routes} from '../routing'
import {asyncValidateArticle} from '../utilities/validation'
import {getSubItems} from '../utilities/getSubItems'
import ReviewArticle from '../components/AddArticlePage/reviewArticleModal'
import { XMLSerializer, DOMParser } from 'xmldom'
import loadArticle from '../components/AddArticlePage/methods/loadArticle'
import save from '../components/AddArticlePage/methods/save'





const mapStateToProps = (state, props) => {
  return ({
    publication: state.publications[props.routeParams.pubDoi],
    reduxForm: state.reduxForm,
    crossmarkPrefixes: state.login['crossmark-prefixes'],
    reduxCart: state.cart
  })
}

const mapDispatchToProps = dispatch => bindActionCreators({
  reduxControlModal: controlModal,
  reduxEditForm: editForm,
  reduxDeleteCard: deleteCard,
  reduxClearForm: clearForm,
  asyncGetPublications: getPublications,
  reduxCartUpdate: cartUpdate
}, dispatch)


@connect(mapStateToProps, mapDispatchToProps)
export default class AddArticlePage extends Component {

  static propTypes = {
    reduxControlModal: is.func.isRequired,
    reduxEditForm: is.func.isRequired,
    reduxClearForm: is.func.isRequired,
    reduxDeleteCard: is.func.isRequired,
    reduxCartUpdate: is.func.isRequired,
    asyncGetPublications: is.func.isRequired,

    routeParams: is.shape({
      pubDoi: is.string.isRequired,
      articleDoi: is.string
    }).isRequired,

    location: is.shape({
      state: is.shape({
        duplicateFrom: is.string
      })
    }).isRequired,

    reduxForm: is.object.isRequired,
    publication: is.object,
    crossmarkPrefixes: is.array.isRequired,
    reduxCart: is.array.isRequired
  }

  constructor (props) {
    super(props)
    const {duplicateFrom, dupIssueDoi, dupIssueTitle} = props.location.state || {}
    const ownerPrefix = props.routeParams.pubDoi.split('/')[0];
    const editArticleDoi = props.routeParams.articleDoi || (props.location.state && props.location.state.duplicateFrom)
    const isDuplicate = !!duplicateFrom
    const issueId = props.routeParams.issueId || dupIssueDoi || dupIssueTitle
    const issueDoi = issueId && (issueId.split('/')[0] === ownerPrefix) ? issueId : undefined
    const issueTitle = issueId && !issueDoi ? JSON.parse(issueId) : undefined
    this.state = {
      ...defaultState,
      publication: props.publication,
      publicationMetaData: {},
      publicationXml: '',
      issue: undefined,
      editArticleDoi,
      issueDoi,
      issueTitle,
      mode: editArticleDoi ? 'edit' : 'add',
      isDuplicate,
      ownerPrefix,
      crossmark: props.crossmarkPrefixes.indexOf(ownerPrefix) !== -1,
      crossmarkCards: {},
      version: '1',
      errorMessages: [],
      focusedInput: '',
      activeCalendar: '',
      scrollToTopButton: false,
      scrollTimeout: null
    }
    this.state.article.doi = ownerPrefix + '/'
  }


  componentDidMount () {
    document.addEventListener('scroll', this.scrollHandler, false)
    loadArticle.call(this)
  }

  save = save.bind(this)


  componentWillReceiveProps (nextProps) {
    if (this.props.crossmarkPrefixes !== nextProps.crossmarkPrefixes) {
      this.setState({
        crossmark: nextProps.crossmarkPrefixes.indexOf(this.state.ownerPrefix) !== -1
      })
    }
  }


  validation = async (data = this.state, reduxForm = this.props.reduxForm, doiDisabled = this.state.doiDisabled) => {
    const {
      criticalErrors,
      warnings,
      licenses,
      contributors,
      relatedItems,
      newReduxForm,
    } = await asyncValidateArticle(data, reduxForm, this.state.ownerPrefix, doiDisabled)

    const validatedPayload = {
      validating: true,
      mode: 'edit',
      error: false,
      errors: {...criticalErrors, ...warnings},
      criticalErrors: criticalErrors,
      license: licenses.length ? licenses : defaultState.license,
      contributors: contributors.length ? contributors : defaultState.contributors,
      relatedItems: relatedItems.length ? relatedItems : defaultState.relatedItems,
      openItems: {
        Contributors: !!contributors.length,
        Funding: !!getSubItems(data.funding).length,
        Licenses: !!licenses.length || !!data.article.freetolicense,
        relatedItems: !!relatedItems.length,
        addInfo: !!getSubItems(data.addInfo).length,
        references: !!data.references.length
      }
    }
    let valid = true

    for(const key in warnings) {
      if (warnings[key]) {
        validatedPayload.error = true
      }
    }

    for(const key in criticalErrors) {
      if(criticalErrors[key]) {
        validatedPayload.error = true
        valid = false
      }
    }

    validatedPayload.errorMessages = this.errorUtility.onValidate(validatedPayload.errors, validatedPayload.contributors, validatedPayload.license, validatedPayload.relatedItems, newReduxForm)

    if(newReduxForm && newReduxForm.size) {
      const keyPath = []
      this.props.reduxEditForm(keyPath, newReduxForm)
    }

    return {valid, validatedPayload}
  }


  validate = async () => {
    if(this.state.mode === 'edit') {
      const {validatedPayload} = await this.validation()
      this.setState(validatedPayload)
    }
  }


  errorUtility = {
    errorIndicators: [],

    activeIndicator: -1,

    openingSubItem: false,
    subItemIndex: "0",

    stickyErrorMounted: false,

    saveRef: (activeErrors, indicatorErrors, node, subItem, subItemIndex, openSubItem) => {
      if(node){
        this.errorUtility.errorIndicators.push({
          activeErrors,
          indicatorErrors,
          node,
          subItem,
          subItemIndex,
          openSubItem
        })

        if(node.id === 'errorBubble') {
          this.errorUtility.activeIndicator = this.errorUtility.errorIndicators.length - 1
        }
      }
    },

    setErrorMessages: (setErrors, allErrors = this.state.errors) => {
      const filteredErrorMessage = setErrors.filter((error)=>{
        return allErrors[error]
      })

      this.setState({errorMessages: filteredErrorMessage})
    },

    onValidate: (newValidationErrors, contributors, license, relatedItems, newReduxForm) => {
      if(!this.state.errorMessages.length) {
        return []
      }

      const {errorIndicators, activeIndicator} = this.errorUtility
      const activeIndicatorObj = errorIndicators[activeIndicator]
      const trackedIndicatorErrors = activeIndicatorObj ? activeIndicatorObj.indicatorErrors : []

      let newErrorMessages
      const {subItem, subItemIndex} = activeIndicatorObj || {}

      if(subItem) {
        const subItemErrors = {
          contributor: contributors,
          license: license,
          relatedItems: relatedItems,
        }

        let allErrors
        if(cardNamesArray.indexOf(subItem) > -1) {
          //Is crossmark subItem
          allErrors = newReduxForm.getIn([subItem, subItemIndex, 'errors'])

        } else {
          allErrors = subItemErrors[subItem][subItemIndex].errors
        }

        newErrorMessages = this.state.errorMessages.filter((error) => {
          return allErrors[error]
        })

        if(!newErrorMessages.length) {
          newErrorMessages = trackedIndicatorErrors.filter((error) => {
            return allErrors[error]
          })
        }


      } else {
        newErrorMessages = this.state.errorMessages.filter((error) => {
          return newValidationErrors[error]
        })

        if(!newErrorMessages.length) {
          newErrorMessages = trackedIndicatorErrors.filter((error)=>{
            return newValidationErrors[error]
          })
        }

        this.errorUtility.subItemIndex = "0"
      }

      if(!newErrorMessages.length) {
        const indicatorBelow = errorIndicators[activeIndicator + 1]
        const indicatorAbove = errorIndicators[activeIndicator - 1]

        if(indicatorBelow) {
          newErrorMessages = indicatorBelow.activeErrors
          this.errorUtility.subItemIndex = indicatorBelow.subItemIndex || "0"
        } else if(indicatorAbove) {
          newErrorMessages = indicatorAbove.activeErrors
          this.errorUtility.subItemIndex = indicatorAbove.subItemIndex || "0"
        }
      }

      return newErrorMessages
    },

    assignStickyErrorRefresh: (func) => {
      this.errorUtility.stickyErrorMounted = true
      this.errorUtility.refreshTask = func
    },

    refreshStickyError: () => {
      if(this.errorUtility.stickyErrorMounted && typeof this.errorUtility.refreshTask === 'function') {
        return finishUpdate().then(()=>this.errorUtility.refreshTask())
      }
    }
  }


  tooltipUtility = {

    getFocusedInput: () => this.state.focusedInput,

    tooltipMounted: false,

    assignRefreshTask: (func) => {
      this.tooltipUtility.tooltipMounted = true
      this.tooltipUtility.refreshTask = func
    },

    refresh: (param) => {
      if(
        this.tooltipUtility.tooltipMounted &&
        typeof this.tooltipUtility.refreshTask === 'function'
      ) {
        return finishUpdate().then(()=>this.tooltipUtility.refreshTask(param))
      }
    },

    assignFocus: (inputId, tooltip) => {
      this.setState({focusedInput: inputId})
      return this.tooltipUtility.refresh(tooltip)
    }
  }


  componentDidUpdate(prevProps, prevState) {
    //Select first error if first validation
    if(
      this.state.validating &&
      this.state.error && !prevState.error
    ) {
      try {
        this.errorUtility.setErrorMessages(this.errorUtility.errorIndicators[0].activeErrors)
      } catch (e) {}
    }

    this.errorUtility.refreshStickyError()

    this.tooltipUtility.refresh()

    this.state.validating = false
    this.state.saving = false
    this.state.openSubItems = false
  }


  openReviewArticleModal = () => {
    this.props.reduxControlModal({
      showModal: true,
      title:
        <div className='innerTitleHolder'>
          <div className='innterTitleHolderIcon'>
            <img src={`${routes.images}/ReviewArticle/Asset_Icons_White_Review.svg`} />
          </div>
          <div className='innerTitleHolderText'>
            {this.state.article.title}
          </div>
        </div>,
      style: 'defaultModal reviewModal',
      Component: ReviewArticle,
      props: {
        submit: this.addToCart,
        reviewData: this.state,
        publication: this.state.publication,
        publicationMetaData: this.state.publicationMetaData,
        issue: this.state.issue
      }
    })
  }


  handleChange = async (e) => {
    this.setState({
      article: {
        ...this.state.article,
        [e.target.name]: e.target.value
      }
    })
  }


  boundSetState = (...args) => { this.setState(...args) }


  toggleFields = () => {
    this.setState({
      showOptionalTitleData: !this.state.showOptionalTitleData
    })
  }


  addSection = (section) => {
    this.setState({
      [section]: [...this.state[section], defaultState[section][0]]
    })
  }


  removeSection = (section, index) => {
    this.setState({
      [section]: do {
        const newArray = [...this.state[section]]
        newArray.splice(index, 1)
        newArray
      }
    })
  }


  addToCart = () => {
    const addToCart = true
    this.save(addToCart)
  }


  back = () => {
    browserHistory.push(`${routes.publications}/${encodeURIComponent(this.state.publication.message.doi)}`)
  }


  crossmarkUtility = {

    addCrossmarkCard: (selection) => {
      this.setState({
        crossmarkCards: {
          ...this.state.crossmarkCards,
          [selection]: this.state.crossmarkCards[selection] ? this.state.crossmarkCards[selection] + 1 : 1
        }
      })
    },

    removeCrossmarkCard: (selection) => {
      const newState = {...this.state.crossmarkCards}
      delete newState[selection]
      this.props.reduxDeleteCard([selection])
      this.setState({
        crossmarkCards: newState
      })
    }
  }


  componentWillUnmount () {
    document.removeEventListener('scroll', this.scrollHandler, false)
    this.props.reduxClearForm();
    clearTimeout(this.state.scrollTimer)
  }


  calendarHandler = (name, dateObj, subItem = 'article', subItemIndex) => {
    if(dateObj) {
      const stateName = this.state.activeCalendar.split('-')[0]
      const datePayload = dateObj ? {
        [`${stateName}Year`]: dateObj.year,
        [`${stateName}Month`]: dateObj.month,
        [`${stateName}Day`]: dateObj.day
      } : {}

      const subItemPayload = subItemIndex ? this.state[subItem].map((item, i) => {
        if(i === Number(subItemIndex)) {
          return {
            ...item,
            ...datePayload
          }
        } else {
          return item
        }
      })
        : {...this.state.article, ...datePayload}

      this.setState({
        activeCalendar: name,
        [subItem]: subItemPayload
      }, () => this.validate())

    } else {
      this.setState({activeCalendar: name})
    }
  }


  scrollHandler = () => {
    if(this.state.scrollTimer) {
      clearTimeout(this.state.scrollTimer)
    }
    this.state.scrollTimer = setTimeout(() => {
      if(window.pageYOffset > 240 && !this.state.scrollToTopButton) {
        this.setState({scrollToTopButton: true})
      }

      if(window.pageYOffset < 240 && this.state.scrollToTopButton) {
        this.setState({scrollToTopButton: false})
      }
    }, 80)
  }


  scrollToTop = () => {
    $('html, body').animate({
      scrollTop: 0
    }, 1000);
  }


  render () {
    this.errorUtility.errorIndicators = []  //Saving refs of any errorIndicators rendered so need to clear it before each render
    this.errorUtility.openingSubItem = false
    return (
      <div className='addArticles'>

        <AddArticleView
          back={this.back}
          addToCart={this.addToCart}
          validate={this.validate}
          save={this.save}
          openReviewArticleModal={this.openReviewArticleModal}
          handleChange={this.handleChange}
          toggleFields={this.toggleFields}
          boundSetState={this.boundSetState}
          removeSection={this.removeSection}
          addSection={this.addSection}
          reduxDeleteCard={this.props.reduxDeleteCard}
          reduxForm={this.props.reduxForm}
          errorUtility={this.errorUtility}
          crossmarkUtility={this.crossmarkUtility}
          tooltipUtility={this.tooltipUtility}
          calendarHandler={this.calendarHandler}
          {...this.state}
        />

        <div className="rightBar">
          {this.state.scrollToTopButton &&
            <div className="scrollToTopButton" onClick={this.scrollToTop}>
              <img className="scrollToTopChevron" src={`${routes.images}/AddArticle/Asset_Icons_White_Chevron.svg`}/>
            </div>}
        </div>
      </div>
    )
  }
}

