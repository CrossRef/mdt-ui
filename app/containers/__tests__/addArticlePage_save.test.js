import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
import sinon from 'sinon'
import {Map} from 'immutable'

import * as api from '../../actions/api'
import {AddArticlePage} from '../addArticlePage'
import defaultArticle from '../../__testUtils/dummyRecords/defaultArticle'
import fullFormArticle from '../../__testUtils/dummyRecords/fullFormArticle'
import recordRegistry from '../../__testUtils/dummyRecords/recordRegistry'




jest.mock("../../actions/api")


const sandbox = sinon.createSandbox()
sandbox.stub(api, 'getItem')
sandbox.stub(api, 'submitItem')


const commonProps = {
  reduxControlModal: () => {},
  reduxEditForm: () => {},
  reduxClearForm: () => {},
  reduxDeleteCard: () => {},
  reduxCartUpdate: () => {},
  asyncGetPublications: () => {},
  reduxForm: new Map,
  reduxCart: [],
  showSection: {},
  crossmarkPrefixes: []
}



function testSave (testName, {pubDoi, editArticleDoi, duplicateArticleDoi}) {
  let component
  let instance
  let save

  beforeEach(async () => {
    const articleProps = {
      ...commonProps,
      routeParams: {pubDoi: pubDoi},
      location: {}
    }

    if(editArticleDoi) {
      articleProps.routeParams.articleDoi = editArticleDoi
    }

    if(duplicateArticleDoi) {
      articleProps.location = {state: {duplicateFrom: duplicateArticleDoi}}
    }


    sandbox.resetHistory()
    api.getItem.callsFake( id => recordRegistry[id].toJS())

    component = mount(
      <AddArticlePage {...articleProps}/>
    )

    instance = component.instance()

  })

  test(testName, async () => {
    if(!editArticleDoi) {
      await component.setState({
        article: {...instance.state.article, title: 'test save', doi: '10.5555/jesttestsave'}
      })

      api.getItem.callsFake( id => {
        throw "Mock doi dupe check"
      })
    }

    await instance.save()

    const savedJson = api.submitItem.args[0][0]
    delete savedJson.message.contains[0].date
    expect(savedJson).toMatchSnapshot()
  })
}


describe('save', ()=>testSave('new article', {pubDoi: defaultArticle.publicationDoi}))

describe('save', ()=>testSave('duplicate defaultArticle', {pubDoi: defaultArticle.publicationDoi, duplicateArticleDoi: defaultArticle.articleDoi}))

describe('save', ()=>testSave('duplicate fullFormArticle', {pubDoi: fullFormArticle.publicationDoi, duplicateArticleDoi: fullFormArticle.articleDoi}))

describe('save', ()=>testSave('edit defaultArticle', {pubDoi: defaultArticle.publicationDoi, editArticleDoi: defaultArticle.articleDoi}))

describe('save', ()=>testSave('edit fullFormArticle', {pubDoi: fullFormArticle.publicationDoi, editArticleDoi: fullFormArticle.articleDoi}))