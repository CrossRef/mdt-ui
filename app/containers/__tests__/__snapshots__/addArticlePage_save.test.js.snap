// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save duplicate defaultArticle 1`] = `
Object {
  "message": Object {
    "contains": Array [
      Object {
        "content": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><crossref xmlns=\\"http://www.crossref.org/xschema/1.1\\"><journal><journal_metadata language=\\"af\\"><full_title>test new system 2</full_title><abbrev_title>arhaerh</abbrev_title><archive_locations><archive name=\\"LOCKSS\\"/></archive_locations><doi_data><doi>10.5555/13g14g14h12h2h24h2</doi><resource>http://afhaehaerhaerh.com</resource></doi_data></journal_metadata><journal_article><titles><title>test save</title></titles><doi_data><doi>10.5555/jesttestsave</doi><resource>http://</resource></doi_data></journal_article></journal></crossref>",
        "deposit-timestamp": null,
        "doi": "10.5555/jesttestsave",
        "mdt-version": "2",
        "owner-prefix": "10.5555",
        "status": "draft",
        "title": Object {
          "title": "test save",
        },
        "type": "article",
      },
    ],
    "date": "2018-01-17",
    "deposit-timestamp": "20180117154727430",
    "doi": "10.5555/defaultPub",
    "mdt-version": "0",
    "owner-prefix": "10.5555",
    "state": Object {},
    "status": "accepted",
    "title": Object {
      "title": "test new system 2",
    },
    "type": "Publication",
  },
}
`;

exports[`save duplicate fullFormArticle 1`] = `
Object {
  "message": Object {
    "contains": Array [
      Object {
        "content": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><crossref xmlns=\\"http://www.crossref.org/xschema/1.1\\"><journal><journal_metadata><full_title>Test dummy data</full_title><doi_data><doi>10.5555/dummypublication</doi><resource>http://dummyurl.com</resource></doi_data></journal_metadata><journal_article language=\\"ab\\"><titles><title>test save</title></titles><contributors><person_name sequence=\\"first\\" contributor_role=\\"chair\\"><given_name>Bob</given_name><surname>Sagat</surname><suffix>Jr.</suffix><affiliation>Full House</affiliation><ORCID>https://orcid.org/0000-0002-3843-3472</ORCID></person_name><organization sequence=\\"additional\\" contributor_role=\\"editor\\">Awesome Group</organization></contributors><jats:abstract xmlns:jats=\\"http://www.ncbi.nlm.nih.gov/JATS1\\"><jats:p>Lorem ipsum

Donec hendrerit.</jats:p></jats:abstract><publication_date media_type=\\"online\\"><month>12</month><day>21</day><year>2016</year></publication_date><publication_date media_type=\\"print\\"><month>1</month><day>5</day><year>2017</year></publication_date><pages><first_page>1</first_page><last_page>5</last_page></pages><publisher_item><item_number item_number_type=\\"article_number\\">3D9324F1-16B1-11D7- 8645000102C</item_number></publisher_item><fr:program xmlns:fr=\\"http://www.crossref.org/fundref.xsd\\"><fr:assertion name=\\"fundgroup\\"><fr:assertion name=\\"funder_name\\">Southern California Clinical and Translational Science Institute<fr:assertion name=\\"funder_identifier\\">http://dx.doi.org/10.13039/100009410</fr:assertion></fr:assertion><fr:assertion name=\\"award_number\\">12</fr:assertion><fr:assertion name=\\"award_number\\">5</fr:assertion></fr:assertion><fr:assertion name=\\"fundgroup\\"><fr:assertion name=\\"funder_name\\">University of Worcester<fr:assertion name=\\"funder_identifier\\">http://dx.doi.org/10.13039/100010022</fr:assertion></fr:assertion><fr:assertion name=\\"award_number\\">4</fr:assertion></fr:assertion></fr:program><ai:program name=\\"AccessIndicators\\" xmlns:ai=\\"http://www.crossref.org/AccessIndicators.xsd\\"><ai:free_to_read start_date=\\"2016-02-20\\"/><ai:license_ref applies_to=\\"tdm\\" start_date=\\"2016-02-20\\">http://adfhaerhaerh.com</ai:license_ref><ai:license_ref applies_to=\\"vor\\" start_date=\\"2016-11-27\\">http://aeraerhaerhaerh.com</ai:license_ref></ai:program><program name=\\"relations\\" xmlns=\\"http://www.crossref.org/relations.xsd\\"><related_item><description>aehraerh</description><inter_work_relation relationship-type=\\"isCommentOn\\" identifier-type=\\"isbn\\">awerh</inter_work_relation></related_item><related_item><description>afhaerh</description><inter_work_relation relationship-type=\\"isReviewOf\\" identifier-type=\\"issn\\">vadfbadfb</inter_work_relation></related_item></program><archive_locations><archive name=\\"LOCKSS\\"/></archive_locations><doi_data><doi>10.5555/jesttestsave</doi><resource>http://aarwhaerh.com</resource><collection property=\\"crawler-based\\"><item crawler=\\"iParadigms\\"><resource>http://ararhareha.com</resource></item></collection></doi_data><citation_list><citation key=\\"ref0\\"><doi>10.1038/nrd2518</doi><unstructured_citation>Lagerstrom MC, Schioth HB. Structural diversity of G protein-coupled receptors and significance for drug discovery. Nat Rev Drug Discov. 2008;7(4):339–57. Epub 2008/04/03. pmid:18382464</unstructured_citation></citation><citation key=\\"ref1\\"><doi>10.1124/mol.63.6.1256</doi><unstructured_citation>Fredriksson R, Lagerstrom MC, Lundin LG, Schioth HB. The G-protein-coupled receptors in the human genome form five main families. Phylogenetic analysis, paralogon groups, and fingerprints. Mol Pharmacol. 2003;63(6):1256–72. pmid:12761335</unstructured_citation></citation><citation key=\\"ref2\\"><doi>10.1016/j.pharmthera.2013.11.004</doi><unstructured_citation>Foster SR, Roura E, Thomas WG. Extrasensory perception: odorant and taste receptors beyond the nose and mouth. Pharmacol Ther. 2014;142(1):41–61. pmid:24280065</unstructured_citation></citation><citation key=\\"ref3\\"><unstructured_citation>Thomas JH, Robertson HM. The Caenorhabditis chemoreceptor gene families. BMC Biol. 2008;6:42. pmid:18837995</unstructured_citation></citation><citation key=\\"ref4\\"><unstructured_citation>The-C.elegans-Sequencing-Consortium. Genome sequence of the nematode C. elegans: a platform for investigating biology. Science. 1998;282(5396):2012–8. pmid:9851916</unstructured_citation></citation></citation_list></journal_article></journal></crossref>",
        "deposit-timestamp": "20180131104528685",
        "doi": "10.5555/jesttestsave",
        "mdt-version": "1",
        "owner-prefix": "10.5555",
        "state": Object {
          "archiveLocation": "LOCKSS",
          "freetolicense": "yes",
        },
        "status": "draft",
        "title": Object {
          "title": "test save",
        },
        "type": "article",
      },
    ],
    "date": "2018-01-30",
    "deposit-timestamp": "20180131104528685",
    "doi": "10.5555/dummypublication",
    "mdt-version": "0",
    "owner-prefix": "10.5555",
    "state": Object {},
    "status": "accepted",
    "title": Object {
      "title": "Test dummy data",
    },
    "type": "Publication",
  },
}
`;

exports[`save edit defaultArticle 1`] = `
Object {
  "message": Object {
    "contains": Array [
      Object {
        "content": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><crossref xmlns=\\"http://www.crossref.org/xschema/1.1\\"><journal><journal_metadata language=\\"af\\"><full_title>test new system 2</full_title><abbrev_title>arhaerh</abbrev_title><archive_locations><archive name=\\"LOCKSS\\"/></archive_locations><doi_data><doi>10.5555/13g14g14h12h2h24h2</doi><resource>http://afhaehaerhaerh.com</resource></doi_data></journal_metadata><journal_article><titles><title>afhaerharhtaer</title></titles><doi_data><doi>10.5555/aefbtabangagnanatn</doi><resource>http://</resource></doi_data></journal_article></journal></crossref>",
        "deposit-timestamp": null,
        "doi": "10.5555/aefbtabangagnanatn",
        "mdt-version": "2",
        "owner-prefix": "10.5555",
        "status": "draft",
        "title": Object {
          "title": "afhaerharhtaer",
        },
        "type": "article",
      },
    ],
    "date": "2018-01-17",
    "deposit-timestamp": "20180117154727430",
    "doi": "10.5555/defaultPub",
    "mdt-version": "0",
    "owner-prefix": "10.5555",
    "state": Object {},
    "status": "accepted",
    "title": Object {
      "title": "test new system 2",
    },
    "type": "Publication",
  },
}
`;

exports[`save edit fullFormArticle 1`] = `
Object {
  "message": Object {
    "contains": Array [
      Object {
        "content": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><crossref xmlns=\\"http://www.crossref.org/xschema/1.1\\"><journal><journal_metadata><full_title>Test dummy data</full_title><doi_data><doi>10.5555/dummypublication</doi><resource>http://dummyurl.com</resource></doi_data></journal_metadata><journal_article language=\\"ab\\"><titles><title>Article Under Issue Dummy Data</title></titles><contributors><person_name sequence=\\"first\\" contributor_role=\\"chair\\"><given_name>Bob</given_name><surname>Sagat</surname><suffix>Jr.</suffix><affiliation>Full House</affiliation><ORCID>https://orcid.org/0000-0002-3843-3472</ORCID></person_name><organization sequence=\\"additional\\" contributor_role=\\"editor\\">Awesome Group</organization></contributors><jats:abstract xmlns:jats=\\"http://www.ncbi.nlm.nih.gov/JATS1\\"><jats:p>Lorem ipsum

Donec hendrerit.</jats:p></jats:abstract><publication_date media_type=\\"online\\"><month>12</month><day>21</day><year>2016</year></publication_date><publication_date media_type=\\"print\\"><month>1</month><day>5</day><year>2017</year></publication_date><pages><first_page>1</first_page><last_page>5</last_page></pages><publisher_item><item_number item_number_type=\\"article_number\\">3D9324F1-16B1-11D7- 8645000102C</item_number></publisher_item><fr:program xmlns:fr=\\"http://www.crossref.org/fundref.xsd\\"><fr:assertion name=\\"fundgroup\\"><fr:assertion name=\\"funder_name\\">Southern California Clinical and Translational Science Institute<fr:assertion name=\\"funder_identifier\\">http://dx.doi.org/10.13039/100009410</fr:assertion></fr:assertion><fr:assertion name=\\"award_number\\">12</fr:assertion><fr:assertion name=\\"award_number\\">5</fr:assertion></fr:assertion><fr:assertion name=\\"fundgroup\\"><fr:assertion name=\\"funder_name\\">University of Worcester<fr:assertion name=\\"funder_identifier\\">http://dx.doi.org/10.13039/100010022</fr:assertion></fr:assertion><fr:assertion name=\\"award_number\\">4</fr:assertion></fr:assertion></fr:program><ai:program name=\\"AccessIndicators\\" xmlns:ai=\\"http://www.crossref.org/AccessIndicators.xsd\\"><ai:free_to_read start_date=\\"2016-02-20\\"/><ai:license_ref applies_to=\\"tdm\\" start_date=\\"2016-02-20\\">http://adfhaerhaerh.com</ai:license_ref><ai:license_ref applies_to=\\"vor\\" start_date=\\"2016-11-27\\">http://aeraerhaerhaerh.com</ai:license_ref></ai:program><program name=\\"relations\\" xmlns=\\"http://www.crossref.org/relations.xsd\\"><related_item><description>aehraerh</description><inter_work_relation relationship-type=\\"isCommentOn\\" identifier-type=\\"isbn\\">awerh</inter_work_relation></related_item><related_item><description>afhaerh</description><inter_work_relation relationship-type=\\"isReviewOf\\" identifier-type=\\"issn\\">vadfbadfb</inter_work_relation></related_item></program><archive_locations><archive name=\\"LOCKSS\\"/></archive_locations><doi_data><doi>10.5555/artundissdummydata</doi><resource>http://aarwhaerh.com</resource><collection property=\\"crawler-based\\"><item crawler=\\"iParadigms\\"><resource>http://ararhareha.com</resource></item></collection></doi_data><citation_list><citation key=\\"ref0\\"><doi>10.1038/nrd2518</doi><unstructured_citation>Lagerstrom MC, Schioth HB. Structural diversity of G protein-coupled receptors and significance for drug discovery. Nat Rev Drug Discov. 2008;7(4):339–57. Epub 2008/04/03. pmid:18382464</unstructured_citation></citation><citation key=\\"ref1\\"><doi>10.1124/mol.63.6.1256</doi><unstructured_citation>Fredriksson R, Lagerstrom MC, Lundin LG, Schioth HB. The G-protein-coupled receptors in the human genome form five main families. Phylogenetic analysis, paralogon groups, and fingerprints. Mol Pharmacol. 2003;63(6):1256–72. pmid:12761335</unstructured_citation></citation><citation key=\\"ref2\\"><doi>10.1016/j.pharmthera.2013.11.004</doi><unstructured_citation>Foster SR, Roura E, Thomas WG. Extrasensory perception: odorant and taste receptors beyond the nose and mouth. Pharmacol Ther. 2014;142(1):41–61. pmid:24280065</unstructured_citation></citation><citation key=\\"ref3\\"><unstructured_citation>Thomas JH, Robertson HM. The Caenorhabditis chemoreceptor gene families. BMC Biol. 2008;6:42. pmid:18837995</unstructured_citation></citation><citation key=\\"ref4\\"><unstructured_citation>The-C.elegans-Sequencing-Consortium. Genome sequence of the nematode C. elegans: a platform for investigating biology. Science. 1998;282(5396):2012–8. pmid:9851916</unstructured_citation></citation></citation_list></journal_article></journal></crossref>",
        "deposit-timestamp": "20180131104528685",
        "doi": "10.5555/artundissdummydata",
        "mdt-version": "1",
        "owner-prefix": "10.5555",
        "state": Object {
          "archiveLocation": "LOCKSS",
          "freetolicense": "yes",
        },
        "status": "draft",
        "title": Object {
          "title": "Article Under Issue Dummy Data",
        },
        "type": "article",
      },
    ],
    "date": "2018-01-31",
    "deposit-timestamp": "20180131104528685",
    "doi": "10.5555/dummypublication",
    "mdt-version": "0",
    "owner-prefix": "10.5555",
    "state": Object {},
    "status": "accepted",
    "title": Object {
      "title": "Test dummy data",
    },
    "type": "Publication",
  },
}
`;

exports[`save new article 1`] = `
Object {
  "message": Object {
    "contains": Array [
      Object {
        "content": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><crossref xmlns=\\"http://www.crossref.org/xschema/1.1\\"><journal><journal_metadata language=\\"af\\"><full_title>test new system 2</full_title><abbrev_title>arhaerh</abbrev_title><archive_locations><archive name=\\"LOCKSS\\"/></archive_locations><doi_data><doi>10.5555/13g14g14h12h2h24h2</doi><resource>http://afhaehaerhaerh.com</resource></doi_data></journal_metadata><journal_article><titles><title>test save</title></titles><doi_data><doi>10.5555/jesttestsave</doi><resource>http://</resource></doi_data></journal_article></journal></crossref>",
        "deposit-timestamp": null,
        "doi": "10.5555/jesttestsave",
        "mdt-version": "1",
        "owner-prefix": "10.5555",
        "status": "draft",
        "title": Object {
          "title": "test save",
        },
        "type": "article",
      },
    ],
    "date": "2018-01-17",
    "deposit-timestamp": "20180117154727430",
    "doi": "10.5555/defaultPub",
    "mdt-version": "0",
    "owner-prefix": "10.5555",
    "state": Object {},
    "status": "accepted",
    "title": Object {
      "title": "test new system 2",
    },
    "type": "Publication",
  },
}
`;
